### ISSUE TITLE
You description of rainbow tables is wrong

### ISSUE THREAD

**Issue by phish** [crypto101/book#361]
"they’re essentially sorted lists of hash function outputs"

A rainbow table only stores a fraction of the hash outputs (e.g. 1 in 10'000). It uses a clever structure to recreate the missing hashes during cracking. Creating a table takes more effort than just generating all hashes (e.g. 10 times more) because some hashes are generated more than once. So, it is not the case that attackers would build a rainbow table if one system uses a single salt
".. once they knew the value of the salt. An attacker would simply compute a single rainbow table for that salt, and compare the results with the hashed passwords". In this situation you would rather just brute-force the passwords, because building a rainbow table would take multiple times this effort. The effort of building a rainbow table is only worth it if you know you can use the same table to hack different systems (because they all have the same salt or no salt). So using a unique and same salt for all your users is a good protection against rainbow table attacks. It still is a bad practice because when the attacker uses brute force to crack the hashes, they only needs to hash each password candidate once and can then look up if it matches any of the user's hashes. If each user had a different salt, they would have to calculate the hash once for each user.

tl;dr: rainbow tables are not just a sorted list, they use some crazy compression that takes a lot of work to create. Using a single salt is not bad because of rainbow tables but because it makes bruteforcing much easier.

**Comment by lvh** [comment]
Hi. Thanks for your feedback. I'm going to try and pick this apart.

Did you read the rest of the chapter? Because "nobody cares about rainbow tables, they're way too expensive to do, just bruteforce every time" is very much the point of the rest of the chapter.

The section on rainbow tables is there because people think salts are the answer, and they end up with something silly like MD5(salt, pw) which does protect against rainbow tables (as you point
